lib/actions-oclif/device/restart.ts

```diff
@@ -0,0 +1,183 @@
+/**
+ * @license
+ * Copyright 2016-2020 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { flags } from '@oclif/command';
+import type { IArg } from '@oclif/parser/lib/args';
+import Command from '../../command';
+import * as cf from '../../utils/common-flags';
+import { getBalenaSdk, getCliUx, stripIndent } from '../../utils/lazy';
+import { tryAsInteger } from '../../utils/validation';
+import type {
+	BalenaSDK,
+	DeviceWithServiceDetails,
+	CurrentServiceWithCommit,
+} from 'balena-sdk';
+
+interface FlagsDef {
+	help: void;
+	service?: string[];
+}
+
+interface ArgsDef {
+	uuid: string;
+}
+
+export default class DeviceRestartCmd extends Command {
+	public static description = stripIndent`
+		Restart containers on a device.
+
+		Restart containers on a device.
+		If the --service flag is provided, then only those services' containers
+		will be restarted, otherwise all containers on the device will be restarted.
+
+		Note this does not reboot the device, to do so use instead \`balena device reboot\`.
+		`;
+	public static examples = [
+		'$ balena device restart 23c73a1',
+		'$ balena device restart 23c73a1 --service myService',
+		'$ balena device restart 23c73a1 -s myService1 -s myService2',
+	];
+
+	public static args: Array<IArg<any>> = [
+		{
+			name: 'uuid',
+			description: 'the uuid of the device to restart',
+			parse: (dev) => tryAsInteger(dev),
+			required: true,
+		},
+	];
+
+	public static usage = 'device restart <uuid>';
+
+	public static flags: flags.Input<FlagsDef> = {
+		service: flags.string({
+			description: 'the service name(s) to restart',
+			char: 's',
+			multiple: true,
+		}),
+		help: cf.help,
+	};
+
+	public static authenticated = true;
+
+	public async run() {
+		const { args: params, flags: options } = this.parse<FlagsDef, ArgsDef>(
+			DeviceRestartCmd,
+		);
+
+		const balena = getBalenaSdk();
+		const ux = getCliUx();
+
+		ux.action.start('Restarting services', '', { stdout: true });
+		if (options.service) {
+			await this.restartServices(balena, params.uuid, options.service);
+		} else {
+			await this.restartAllServices(balena, params.uuid);
+		}
+		ux.action.stop();
```

Comment:
Better have `action.stop()` in a finally block, so it always runs in case the restart methods throw an error.

Reply:
AFAIK `action.stop()` simply changes the spinner status to`done` (by default) and removes the spinner.  If you don't stop it then it prints a red `!` next the the last status (and error is output under it) which is what we want I think.

Reply:
Interesting, just tested it with an explicit `throw new Error('oops')` line between spinner start/stop, and indeed it shows the red exclamation mark ❗️  _before_ 'oops' is printed to the console. So it must be somehow intercepting thrown errors, in a more "intrusive" way than I expected. Hopefully doesn't conflict with Sentry. It seems to work well though. 👍